test/contract/IteratableOrderSet.spec.ts

```diff
@@ -233,6 +236,110 @@ describe("IterableOrderedOrderSet", function () {
     expect(await set.callStatic.remove(BYTES32_TWO)).to.equal(false);
   });
 
+  describe("insert elements using removed element as reference", () => {
+    it("single element removed", async () => {
+      await set.insert(BYTES32_ONE);
+      await set.insert(BYTES32_TWO);
+      await set.insert(BYTES32_FOUR);
+
+      await set.removeKeepHistory(BYTES32_TWO);
+      expect(
+        await set.callStatic.insertAt(BYTES32_THREE, BYTES32_TWO),
+      ).to.equal(true);
+      await set.insertAt(BYTES32_THREE, BYTES32_TWO);
+
+      const first = await set.first();
+      const second = await set.next(first);
+      const third = await set.next(second);
+      const next_of_third = await set.nextMap(third);
+
+      expect(first).to.equal(BYTES32_ONE);
+      expect(second).to.equal(BYTES32_THREE);
+      expect(third).to.equal(BYTES32_FOUR);
+      expect(next_of_third).to.equal(QUEUE_END);
+    });
+
+    it("two elements removed, backtrack once", async () => {
+      await set.insert(BYTES32_ONE);
+      await set.insert(BYTES32_TWO);
+      await set.insert(BYTES32_THREE);
+      await set.insert(BYTES32_FIVE);
+
+      await set.removeKeepHistory(BYTES32_TWO);
+      expect(await set.prevMap(BYTES32_TWO)).to.equal(BYTES32_ONE);
+      expect(await set.nextMap(BYTES32_TWO)).to.equal(ethers.constants.Zero);
+      await set.removeKeepHistory(BYTES32_THREE);
+      expect(await set.prevMap(BYTES32_THREE)).to.equal(BYTES32_ONE);
+      expect(await set.nextMap(BYTES32_THREE)).to.equal(ethers.constants.Zero);
+      await set.insertAt(BYTES32_FOUR, BYTES32_TWO);
+
+      const first = await set.first();
+      const second = await set.next(first);
+      const third = await set.next(second);
+      const next_of_third = await set.nextMap(third);
+
+      expect(first).to.equal(BYTES32_ONE);
+      expect(second).to.equal(BYTES32_FOUR);
+      expect(third).to.equal(BYTES32_FIVE);
+      expect(next_of_third).to.equal(QUEUE_END);
+    });
+
+    it("two elements removed, backtrack twice", async () => {
+      await set.insert(BYTES32_ONE);
+      await set.insert(BYTES32_TWO);
+      await set.insert(BYTES32_THREE);
+      await set.insert(BYTES32_FIVE);
+
+      await set.removeKeepHistory(BYTES32_THREE);
+      expect(await set.prevMap(BYTES32_THREE)).to.equal(BYTES32_TWO);
+      expect(await set.nextMap(BYTES32_THREE)).to.equal(ethers.constants.Zero);
+      await set.removeKeepHistory(BYTES32_TWO);
+      expect(await set.prevMap(BYTES32_TWO)).to.equal(BYTES32_ONE);
+      expect(await set.nextMap(BYTES32_TWO)).to.equal(ethers.constants.Zero);
+      await set.insertAt(BYTES32_FOUR, BYTES32_TWO);
```

Comment:
for "backtracking" twice, you want to use Bytes32_Three, right?

Reply:
Very nice catch, fixing it!